common/utils.c

```diff
@@ -416,3 +416,106 @@ TDNFYesOrNo(
     return dwError;
 }
 
+uint32_t TDNFUriIsRemote(
+    const char* pszKeyUrl,
+    int *pnRemote
+)
+{
+    uint32_t dwError = 0;
+    const char *szRemotes[] = {"http://", "https://", "ftp://",
+                               "ftps://", NULL};
+    int i = 0;
+
+    if(!pnRemote || IsNullOrEmptyString(pszKeyUrl))
+    {
+      dwError = ERROR_TDNF_INVALID_PARAMETER;
+      BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    *pnRemote = 0;
+    for(i = 0; szRemotes[i]; i++) {
+        if (strncasecmp(pszKeyUrl, szRemotes[i], strlen(szRemotes[i])) == 0) {
+            *pnRemote = 1;
+            break;
+        }
+    }
+    if (szRemotes[i] == NULL && strncasecmp(pszKeyUrl, "file://", 7) != 0) {
+       dwError = ERROR_TDNF_KEYURL_INVALID;
+       BAIL_ON_TDNF_ERROR(dwError);
+    }
+cleanup:
+    return dwError;
+
+error:
+    goto cleanup;
+}
+
+uint32_t TDNFFileNameFromUri(
+    const char* pszKeyUrl,
+    char** ppszFile)
+{
+    uint32_t dwError = 0;
+    const char* pszPath = NULL;
+    char *pszFile = NULL;
+    size_t nOffset;
+    const char *szPrefixes[] = {"http://", "https://", "ftp://",
+                                "ftps://", "file://", NULL};
+    int i = 0;
+
+    if(IsNullOrEmptyString(pszKeyUrl) || !ppszFile)
+    {
+      dwError = ERROR_TDNF_INVALID_PARAMETER;
+      BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    for(i = 0; szPrefixes[i]; i++) {
+        if (strncasecmp(pszKeyUrl, szPrefixes[i], strlen(szPrefixes[i])) == 0) {
+            nOffset = strlen(szPrefixes[i]);
+            break;
+        }
+    }
+    if (szPrefixes[i] == NULL) {
+       dwError = ERROR_TDNF_KEYURL_INVALID;
+       BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pszPath = pszKeyUrl + nOffset;
+
+    if(!pszPath || *pszPath == '\0')
+    {
+        dwError = ERROR_TDNF_KEYURL_INVALID;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (strchr (pszPath, '#') != NULL)
```

Comment:
This method must be a generic `get_file_name_from_url()`. So '#' is a valid URL char (for anchors) and must be treated accordingly.

So is a '?' (URL params) and other URL defined reserved chars. These are the two common URL additional info strings that I can remember. For starters handling these 2 would be sufficient.

Reply:
I didn't write this code, just moved it from `client/gpgcheck.c`. It's a few years old. I found out that originally the function `g_filename_from_uri()` from glib was used - which would be ideal because it would handle all sorts of corner cases and also unescape the URL, but we moved away from glib. Recent versions of `libcurl` have the function `curl_url_get()` which we could use, see https://curl.haxx.se/libcurl/c/curl_url_get.html. Unfortunately this is not in Photon 3 yet. But since we link with curl anyway, I think we should use that as soon as it's available.

I don't think we have any use case in tdnf where we would accept a `#` in a URL. It's only useful AFAIK for HTML pages, which we don't use. A `?` may be a valid use case (although odd).

Reply:
fyi - glib removal is a conscious decision to reduce install size.

Reply:
Okay.